### PR TITLE
ABI Generation

### PR DESCRIPTION
**Pull request by containerman17**
Generate a JSON representation of all VM actions and their fields.

This includes all commits from https://github.com/ava-labs/hypersdk/pull/1198, so it should be merged after that.

When reviewing the code, please [compare it with the code in PR 1198](https://github.com/ava-labs/hypersdk/compare/containerman/standardize-marshal-function...containerman17/abi), not with the main branch.

### CODE REVIEW

File: chain/dependencies.go [hunk 1]
```diff
@@ -185,6 +185,11 @@ type StateManager interface {
 	MetadataManager
 }
 
+type Marshaler interface {
+	// Marshal encodes an [Action] as bytes.
+	Marshal(p *codec.Packer)
```
Comment (richardpringle) [inline chain/dependencies.go:190]:
```suggestion
type Marshaller interface {
	// Marshal encodes an [Action] as bytes.
	Marshal(p *codec.Packer)
```

File: chain/transaction.go [hunk 2]
```diff
@@ -59,7 +59,11 @@ func (t *Transaction) Digest() ([]byte, error) {
 	p.PackByte(uint8(len(t.Actions)))
 	for _, action := range t.Actions {
 		p.PackByte(action.GetTypeID())
-		action.Marshal(p)
+		if marshaler, ok := action.(Marshaler); ok {
```
Comment (richardpringle) [inline chain/transaction.go:62]:
```suggestion
		if marshaller, ok := action.(Marshaller); ok {
```

File: chain/transaction.go [hunk 3]
```diff
@@ -366,7 +370,11 @@ func (t *Transaction) marshalActions(p *codec.Packer) error {
 	for _, action := range t.Actions {
 		actionID := action.GetTypeID()
 		p.PackByte(actionID)
-		action.Marshal(p)
+		if marshaler, ok := action.(Marshaler); ok {
+			marshaler.Marshal(p)
```
Comment (richardpringle) [inline chain/transaction.go:374]:
```suggestion
		if marshaller, ok := action.(Marshaller); ok {
			marshaller.Marshal(p)
```

**Review by richardpringle — CHANGES_REQUESTED** [review]
Small spelling mistake

**Comment by containerman17** [comment]
Blocked by https://github.com/ava-labs/hypersdk/pull/1198

**Comment by containerman17** [comment]
Closed in favor of https://github.com/ava-labs/hypersdk/pull/1459

**Pull request closed without merging**